feat: add drag n drop functionality to section cards

steff456 opened the pull request:
Resolves `private ref`: [BB-8219](https://tasks.opencraft.com/browse/BB-8219)

This PR,

- [x] Adds an endpoint for reordering the list of sections
- [x] Adds the functionality for drag and drop the sections cards
- [x] Adds new tests for the functionality

**Demo**

https://github.com/open-craft/frontend-app-course-authoring/assets/20992645/af0ee84e-e558-4d42-87a9-841ba0ebe417

**Depends on**

- https://github.com/open-craft/frontend-app-course-authoring/pull/12

**Testing instructions**
- Follow instructions from the readme
- Checkout branch from https://github.com/openedx/edx-platform/pull/33667 and run `make cms-up` in devstack.
- Open any course outline by using this url format: http://localhost:2001/course/course-v1:edX+DemoX+Demo_Course/outline
- Drag and drop sections among them!

**Questions**

- I'm unsure if there's any preference on the style of the draggable objects, in the other front end they are using three dots so the users can understand they are able to drag and drop the cards.
- The same question goes for the `drag` animation, right now it is using the default

steff456 committed:
feat: add drag n drop functionality to section cards

navinkarkera commented:
@steff456 Even though I like that the custom drag-n-drop implementation, we should probably use something like `react-DnD` as the current implementation is bit confusing as their is no indication about the drop position and implementing it by hand will require a lot more effort and time. See this [example](https://react-dnd.github.io/react-dnd/examples/sortable/simple), we want something similar. Also, after dropping an item it should not go back to its original position and wait for the API to complete, instead it should update the UI first and then update the backend via API and if for some reason the API call fails, we display an error message (which is already handled in exception handling).

> I'm unsure if there's any preference on the style of the draggable objects, in the other front end they are using three dots so the users can understand they are able to drag and drop the cards.

We don't need to add the three dots as per the new design. The curson already updates as expected so we don't need any changes in UI.

> The same question goes for the drag animation, right now it is using the default

We can keep the default animation for now from the [example](https://react-dnd.github.io/react-dnd/examples/sortable/simple)

navinkarkera reviewed (changes requested):
@steff456 Even though I like that the custom drag-n-drop implementation, we should probably use something like `react-DnD` as the current implementation is bit confusing as their is no indication about the drop position and implementing it by hand will require a lot more effort and time. See this [example](https://react-dnd.github.io/react-dnd/examples/sortable/simple), we want something similar. Also, after dropping an item it should not go back to its original position and wait for the API to complete, instead it should update the UI first and then update the backend via API and if for some reason the API call fails, we display an error message (which is already handled in exception handling).

> I'm unsure if there's any preference on the style of the draggable objects, in the other front end they are using three dots so the users can understand they are able to drag and drop the cards.

We don't need to add the three dots as per the new design. The curson already updates as expected so we don't need any changes in UI.

> The same question goes for the drag animation, right now it is using the default

We can keep the default animation for now from the [example](https://react-dnd.github.io/react-dnd/examples/sortable/simple)

steff456 commented:
@navinkarkera I'll modify the implementation to use `react-dnd` and ping you once it is ready :)

steff456 committed:
feat: use react-dnd library for drag and drop implementation

steff456 commented:
@navinkarkera I just updated to use `dnd-react` library, I also updated the demo video in the description. With this changes now the UI updates first and then it sends the new order to the backend and now the animation is the same as the example you provided. 

Tomorrow I'll work on adding some tests and fixing the linting issues that appeared in the CI :)

steff456 committed:
style: fix linting issues

navinkarkera committed:
fix: finalize section order on drop instead of hover

navinkarkera commented on src/course-outline/CourseOutline.jsx:
```diff
@@ -69,15 +74,40 @@ const CourseOutline = ({ courseId }) => {
+      for (let i = 0; i < newList.length; i++) {
+        idList.push(newList[i].id);
+      }
+      handleDragNDrop(idList);
```
This is sending multiple put requests on hover, instead we should only send a request on drop.

navinkarkera reviewed (changes requested):
@steff456 Thanks, I just pushed a commit to fix multiple requests on hover issue.

I see ~43 new commits as you seem to have use merge commit instead of rebasing with the base branch. Can you try to create a new branch from the base and cherry pick your and my last commit to cleanup the commit history?

steff456 committed:
fix: prevent same index drag to start request and restore state on error

navinkarkera committed:
fix: restore sectionlist order

navinkarkera commented on src/course-outline/CourseOutline.jsx:
```diff
@@ -69,15 +74,43 @@ const CourseOutline = ({ courseId }) => {
   document.title = getPageHeadTitle(courseName, intl.formatMessage(messages.headingTitle));
+  const [sections, setSections] = useState(sectionsList);
+
+  const initialSections = [...sectionsList];
```
@steff456 There was an issue with restoring section list back to original when the API fails as this line was `const initialSections = [...sections];` which is updated with sections so the restore did not work. I took the liberty to fix this and update sectionListOrder in reducer when the request is successful to handle failures in API after a successful attempt.

steff456 replied:
thanks for helping me with that!

navinkarkera reviewed (changes requested):
@steff456 Almost there. We have one more issue where if you try to select text via dragging the mouse in title while editing it, it drags the whole section. Maybe this issue can help fixing it: https://github.com/react-dnd/react-dnd/issues/335

steff456 committed:
fix: prevent drag event while editing the text

steff456 commented:
@navinkarkera the last commit fixes the text editing issue. Please let me know if there's anything else this PR needs.

steff456 committed:
style: fix linting issues

steff456 committed:
test: fix failing tests

steff456 committed:
test: add missing hooks to SectionCard component in test

steff456 committed:
test: add wrapping to SectionCard test component

navinkarkera commented on src/course-outline/section-card/SectionCard.jsx:
```diff
@@ -96,6 +174,7 @@ const SectionCard = ({
         onEditSubmit={handleEditSubmit}
         isDisabledEditField={savingStatus === RequestStatus.IN_PROGRESS}
         onClickDuplicate={onDuplicateSubmit}
+        className="nodrag"
```
@steff456 It seems to have no effect, we can remove this.

navinkarkera reviewed (changes requested):
@steff456 The functionality now works as expected. We have an issue in firefox where the text inside the form cannot be selected. We can address this in a separate ticket as HTML5 drag n drop has some issues with firefox. For now it looks good.

We should add some tests (probably in `CourseOutline.test.jsx`) to test ordering when it runs successfully as well as when something goes wrong (restoreCallback).

steff456 committed:
test: add tests for checking the API that sets the ordering

steff456 committed:
fix: merge scroll-to-bottom with drag and drop implementations

steff456 committed:
fix: fix linting issues

steff456 commented:
@navinkarkera I just updated this branch with the latest changes of the base branch. I had to do some modifications for the scroll to work with the drag and drop given that both were using `refs` and `useEffect` simultaneously, but everything is working as expected now. The CI failed because of coverage, which I suspect is because of the fork and base branch PR. 

Please let me know if there's anything else needed in this PR for merging!

navinkarkera committed:
fix: handle scrolling with drag-n-drop

navinkarkera committed:
test: update tests

steff456 commented:
Thanks for all the review @navinkarkera! 🎉
